File: frontend/src/services/interviewV2.js
/**
 * Enhanced Interview Service with state management and threading support
 */

const API_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

// Get auth headers
const getAuthHeaders = () => {
  const token = localStorage.getItem('token');
  return {
    'Content-Type': 'application/json',
    ...(token && { 'Authorization': `Bearer ${token}` })
  };
};

/**
 * Start a new interview session
 */
export const startInterviewSession = async (interviewData) => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/start`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        role: interviewData.role,
        company: interviewData.company,
        resume_text: interviewData.resume_text
      }),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to start interview');
    }

    return data;
  } catch (error) {
    console.error('Start interview error:', error);
    throw error;
  }
};

/**
 * Submit an answer for a specific question
 */
export const submitAnswer = async (threadId, questionNumber, answer) => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/submit-answer`, {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({
        thread_id: threadId,
        question_number: questionNumber,
        answer: answer
      }),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to submit answer');
    }

    return data;
  } catch (error) {
    console.error('Submit answer error:', error);
    throw error;
  }
};

/**
 * Get chat history for an interview session
 */
export const getChatHistory = async (threadId) => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/chat-history/${threadId}`, {
      method: 'GET',
      headers: getAuthHeaders(),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to get chat history');
    }

    return data;
  } catch (error) {
    console.error('Get chat history error:', error);
    throw error;
  }
};

/**
 * Get all interview sessions for the current user
 */
export const getUserInterviewSessions = async () => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/sessions`, {
      method: 'GET',
      headers: getAuthHeaders(),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to get interview sessions');
    }

    return data;
  } catch (error) {
    console.error('Get interview sessions error:', error);
    throw error;
  }
};

/**
 * Get status and progress of an interview session
 */
export const getSessionStatus = async (threadId) => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/session/${threadId}/status`, {
      method: 'GET',
      headers: getAuthHeaders(),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to get session status');
    }

    return data;
  } catch (error) {
    console.error('Get session status error:', error);
    throw error;
  }
};

/**
 * Delete an interview session
 */
export const deleteInterviewSession = async (threadId) => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/session/${threadId}`, {
      method: 'DELETE',
      headers: getAuthHeaders(),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to delete interview session');
    }

    return data;
  } catch (error) {
    console.error('Delete interview session error:', error);
    throw error;
  }
};

/**
 * Get user analytics and performance metrics
 */
export const getUserAnalytics = async () => {
  try {
    const response = await fetch(`${API_URL}/interview/v2/analytics`, {
      method: 'GET',
      headers: getAuthHeaders(),
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to get analytics');
    }

    return data;
  } catch (error) {
    console.error('Get analytics error:', error);
    throw error;
  }
};

/**
 * Upload resume and extract text
 */
export const uploadResume = async (file) => {
  try {
    const formData = new FormData();
    formData.append('file', file);

    const token = localStorage.getItem('token');
    const headers = {};
    if (token) {
      headers['Authorization'] = `Bearer ${token}`;
    }

    const response = await fetch(`${API_URL}/interview/upload-resume`, {
      method: 'POST',
      headers: headers,
      body: formData,
    });

    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.detail || data.message || 'Failed to upload resume');
    }

    return data;
  } catch (error) {
    console.error('Upload resume error:', error);
    throw error;
  }
};

/**
 * Format chat messages for display
 */
export const formatChatHistory = (chatHistory) => {
  if (!chatHistory || !chatHistory.messages) {
    return [];
  }

  return chatHistory.messages.map(message => ({
    id: message.id,
    type: message.type,
    role: message.role,
    content: message.content,
    questionNumber: message.question_number,
    marks: message.marks,
    timestamp: new Date(message.created_at),
    metadata: message.metadata
  }));
};

/**
 * Calculate progress percentage
 */
export const calculateProgress = (sessionStatus) => {
  if (!sessionStatus) return 0;
  
  const { answers_submitted = 0, total_questions = 3 } = sessionStatus;
  return Math.round((answers_submitted / total_questions) * 100);
};

/**
 * Get status color for UI
 */
export const getStatusColor = (status) => {
  switch (status) {
    case 'active':
    case 'in_progress':
      return 'text-blue-600';
    case 'completed':
      return 'text-green-600';
    case 'archived':
      return 'text-gray-600';
    default:
      return 'text-gray-600';
  }
};

/**
 * Format score for display
 */
export const formatScore = (score) => {
  if (typeof score !== 'number') return '0.0';
  return score.toFixed(1);
};

File: backend/api/interview_v2.py
"""
Enhanced Interview API with LangGraph state management and threading support.
"""

from fastapi import APIRouter, HTTPException, Depends, status
from sqlalchemy.orm import Session
from typing import List, Dict, Any
import uuid

from database import get_db
from models import (
    User, InterviewSession, ChatMessage,
    InterviewStartRequest, AnswerSubmissionRequest,
    ChatHistoryResponse, InterviewSessionResponse
)
from api.auth import get_current_user
from state_manager import interview_manager

router = APIRouter(prefix="/interview/v2", tags=["interview-v2"])

@router.post("/start", response_model=Dict[str, Any])
async def start_interview_session(
    request: InterviewStartRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Start a new interview session with LangGraph state management.
    """
    try:
        # Create interview session using state manager
        result = interview_manager.create_interview_session(
            user_id=current_user.id,
            role=request.role,
            company=request.company,
            resume_text=request.resume_text
        )
        
        return {
            "message": "Interview session started successfully",
            "thread_id": result["thread_id"],
            "session_id": result["session_id"],
            "questions": result["questions"],
            "status": result["status"],
            "role": request.role,
            "company": request.company
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to start interview session: {str(e)}"
        )

@router.post("/submit-answer", response_model=Dict[str, Any])
async def submit_answer(
    request: AnswerSubmissionRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Submit an answer for a specific question in the interview.
    """
    try:
        # Verify the thread belongs to the current user
        db_session = db.query(InterviewSession).filter(
            InterviewSession.thread_id == request.thread_id,
            InterviewSession.user_id == current_user.id
        ).first()
        
        if not db_session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview session not found or access denied"
            )
        
        # Submit answer using state manager
        result = interview_manager.submit_answer(
            thread_id=request.thread_id,
            question_number=request.question_number,
            answer=request.answer
        )
        
        return {
            "message": "Answer submitted successfully",
            "thread_id": result["thread_id"],
            "question_number": result["question_number"],
            "status": result["status"],
            "current_question": result["current_question"],
            "total_score": result["total_score"],
            "average_score": result["average_score"],
            "completed": result["completed"]
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to submit answer: {str(e)}"
        )

@router.get("/chat-history/{thread_id}")
async def get_chat_history(
    thread_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get complete chat history for an interview session.
    """
    try:
        # Verify the thread belongs to the current user
        db_session = db.query(InterviewSession).filter(
            InterviewSession.thread_id == thread_id,
            InterviewSession.user_id == current_user.id
        ).first()
        
        if not db_session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview session not found or access denied"
            )
        
        # Get chat history using state manager
        chat_history = interview_manager.get_chat_history(thread_id)
        
        return chat_history
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get chat history: {str(e)}"
        )

@router.get("/sessions")
async def get_user_interview_sessions(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get all interview sessions for the current user.
    """
    try:
        sessions = interview_manager.get_user_sessions(current_user.id)
        
        return {
            "message": "Interview sessions retrieved successfully",
            "user_id": current_user.id,
            "sessions": sessions,
            "total_sessions": len(sessions)
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get interview sessions: {str(e)}"
        )

@router.get("/session/{thread_id}/status")
async def get_session_status(
    thread_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get the current status and progress of an interview session.
    """
    try:
        # Verify the thread belongs to the current user
        db_session = db.query(InterviewSession).filter(
            InterviewSession.thread_id == thread_id,
            InterviewSession.user_id == current_user.id
        ).first()
        
        if not db_session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview session not found or access denied"
            )
        
        # Get current state from state manager
        if thread_id in interview_manager.states:
            state = interview_manager.states[thread_id]
            
            return {
                "thread_id": thread_id,
                "session_id": db_session.id,
                "role": db_session.role,
                "company": db_session.company,
                "status": state.get("status", db_session.status),
                "current_question": state.get("current_question", 0),
                "total_questions": len(state.get("questions", [])),
                "answers_submitted": len([a for a in state.get("answers", []) if a]),
                "total_score": state.get("total_score", db_session.total_score or 0.0),
                "average_score": state.get("average_score", db_session.average_score or 0.0),
                "created_at": db_session.created_at.isoformat(),
                "completed_at": db_session.completed_at.isoformat() if db_session.completed_at else None
            }
        else:
            # Fallback to database information
            return {
                "thread_id": thread_id,
                "session_id": db_session.id,
                "role": db_session.role,
                "company": db_session.company,
                "status": db_session.status,
                "total_score": db_session.total_score or 0.0,
                "average_score": db_session.average_score or 0.0,
                "created_at": db_session.created_at.isoformat(),
                "completed_at": db_session.completed_at.isoformat() if db_session.completed_at else None
            }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get session status: {str(e)}"
        )

@router.delete("/session/{thread_id}")
async def delete_interview_session(
    thread_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Delete an interview session and all associated data.
    """
    try:
        # Verify the thread belongs to the current user
        db_session = db.query(InterviewSession).filter(
            InterviewSession.thread_id == thread_id,
            InterviewSession.user_id == current_user.id
        ).first()
        
        if not db_session:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Interview session not found or access denied"
            )
        
        # Delete chat messages
        db.query(ChatMessage).filter(ChatMessage.session_id == db_session.id).delete()
        
        # Delete session
        db.delete(db_session)
        db.commit()
        
        # Remove from state manager memory
        if thread_id in interview_manager.states:
            del interview_manager.states[thread_id]
        
        return {
            "message": "Interview session deleted successfully",
            "thread_id": thread_id
        }
        
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete interview session: {str(e)}"
        )

@router.get("/analytics")
async def get_user_analytics(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """
    Get analytics and performance metrics for the user's interviews.
    """
    try:
        sessions = db.query(InterviewSession).filter(
            InterviewSession.user_id == current_user.id,
            InterviewSession.status == "completed"
        ).all()
        
        if not sessions:
            return {
                "message": "No completed interviews found",
                "total_interviews": 0,
                "analytics": {}
            }
        
        # Calculate analytics
        total_interviews = len(sessions)
        total_scores = [s.total_score for s in sessions if s.total_score]
        average_scores = [s.average_score for s in sessions if s.average_score]
        
        analytics = {
            "total_interviews": total_interviews,
            "average_total_score": sum(total_scores) / len(total_scores) if total_scores else 0,
            "average_score_per_question": sum(average_scores) / len(average_scores) if average_scores else 0,
            "best_performance": max(average_scores) if average_scores else 0,
            "worst_performance": min(average_scores) if average_scores else 0,
            "companies_interviewed": list(set(s.company for s in sessions if s.company)),
            "roles_interviewed": list(set(s.role for s in sessions if s.role)),
            "completion_rate": len([s for s in sessions if s.status == "completed"]) / len(sessions) * 100 if sessions else 0
        }
        
        return {
            "message": "Analytics retrieved successfully",
            "user_id": current_user.id,
            "analytics": analytics
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get analytics: {str(e)}"
        )
#